firebase-segmentation/src/main/java/com/google/firebase/segmentation/CustomInstallationIdCache.java

```diff
@@ -0,0 +1,130 @@
+    // errors.
+    PENDING,
+    // Cache entry is not accepted by Firebase backend.
+    ERROR,
+  }
+
+  private static final String SHARED_PREFS_NAME = "CustomInstallationIdCache";
+
+  private static final String CUSTOM_INSTALLATION_ID_KEY = "Cid";
+  private static final String INSTANCE_ID_KEY = "Iid";
+  private static final String CACHE_STATUS_KEY = "Status";
+
+  private static CustomInstallationIdCache singleton = null;
+  private final Executor ioExecuter;
+  private final SharedPreferences prefs;
+
+  static CustomInstallationIdCache getInstance() {
+    if (singleton == null) {
+      singleton = new CustomInstallationIdCache();
+    }
+    return singleton;
+  }
+
+  private CustomInstallationIdCache() {
+    // Since different FirebaseApp in the same Android application should have the same application
+    // context and same dir path, so that use the context of the default FirebaseApp to create the
+    // shared preferences.
+    prefs =
+        FirebaseApp.getInstance()
+            .getApplicationContext()
+            .getSharedPreferences(SHARED_PREFS_NAME, Context.MODE_PRIVATE);
+
+    ioExecuter = Executors.newFixedThreadPool(2);
+  }
+
+  @Nullable
+  synchronized CustomInstallationIdCacheEntryValue readCacheEntryValue(FirebaseApp firebaseApp) {
+    String cid =
+        prefs.getString(getSharedPreferencesKey(firebaseApp, CUSTOM_INSTALLATION_ID_KEY), null);
+    String iid = prefs.getString(getSharedPreferencesKey(firebaseApp, INSTANCE_ID_KEY), null);
+    int status = prefs.getInt(getSharedPreferencesKey(firebaseApp, CACHE_STATUS_KEY), -1);
+
+    if (Strings.isEmptyOrWhitespace(cid) || Strings.isEmptyOrWhitespace(iid) || status == -1) {
+      return null;
+    }
+
+    return CustomInstallationIdCacheEntryValue.create(cid, iid, CacheStatus.values()[status]);
+  }
+
+  synchronized Task<Boolean> insertOrUpdateCacheEntry(
+      FirebaseApp firebaseApp, CustomInstallationIdCacheEntryValue entryValue) {
+    SharedPreferences.Editor editor = prefs.edit();
+    editor.putString(
+        getSharedPreferencesKey(firebaseApp, CUSTOM_INSTALLATION_ID_KEY),
+        entryValue.getCustomInstallationId());
+    editor.putString(
+        getSharedPreferencesKey(firebaseApp, INSTANCE_ID_KEY), entryValue.getFirebaseInstanceId());
+    editor.putInt(
+        getSharedPreferencesKey(firebaseApp, CACHE_STATUS_KEY),
+        entryValue.getCacheStatus().ordinal());
+    return commitSharedPreferencesEditAsync(editor);
+  }
+
+  synchronized Task<Boolean> clear(FirebaseApp firebaseApp) {
+    SharedPreferences.Editor editor = prefs.edit();
+    editor.remove(getSharedPreferencesKey(firebaseApp, CUSTOM_INSTALLATION_ID_KEY));
+    editor.remove(getSharedPreferencesKey(firebaseApp, INSTANCE_ID_KEY));
+    editor.remove(getSharedPreferencesKey(firebaseApp, CACHE_STATUS_KEY));
+    return commitSharedPreferencesEditAsync(editor);
+  }
+
+  @VisibleForTesting
```

Comment:
Optionally consider using `@RestrictTo(RestrictTo.Scope.Tests)`, as it will also warn you if you try to call it from within this class.